### PR TITLE
Fix ViewPanel layout issue when outer border exists

### PR DESCRIPTION
**Pull request by ddalp**
#3549 
#3488 
#3661 

RNW implementation of ViewPanel sometimes create outer border instead of using the inner border, for example when round corner border is requested. This is causing noticeable layout issue especially when border size is large. RN's yoga layout sets the Top/Left for each children, but Yoga has no notion of outer border(in its view, border is always part of the view instead of parent of view), and its placement logic includes the border size. So we need to compensate the outer border size when ViewPanel is arranging its children, by deducting the border size from the Top and Left.

When testing, I also found children clipping issue when outer border is present, so fix that too.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3660)

### CODE REVIEW

File: vnext/ReactUWP/Views/ViewPanel.cpp [hunk 1]
```diff
@@ -139,6 +139,16 @@ winrt::Size ViewPanel::MeasureOverride(winrt::Size availableSize) {
 }
 
 winrt::Size ViewPanel::ArrangeOverride(winrt::Size finalSize) {
+  // Sometimes we create outerBorder(i.e. when CornerRadius is true) instead of innerBorder,
+  // Yoga has no notion of outerBorder when calculating the child's position, so we
+  // need to make adujustments in arrange for outerborder's thickness.
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/ViewPanel.cpp]:
```suggestion
  // need to make adjustments in arrange for outerborder's thickness.
```

File: vnext/ReactUWP/Views/ViewPanel.cpp [hunk 2]
```diff
@@ -166,8 +176,10 @@ winrt::Size ViewPanel::ArrangeOverride(winrt::Size finalSize) {
     childWidth = std::max<double>(0.0f, childWidth);
     childHeight = std::max<double>(0.0f, childHeight);
 
-    child.Arrange(winrt::Rect(
-        (float)ViewPanel::GetLeft(child), (float)ViewPanel::GetTop(child), (float)childWidth, (float)childHeight));
+    float adjustedLeft = (float)ViewPanel::GetLeft(child) - outerBorderLeft;
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/ViewPanel.cpp]:
```suggestion
    float adjustedLeft = static_cast<float>(ViewPanel::GetLeft(child)) - outerBorderLeft;
```

File: vnext/ReactUWP/Views/ViewPanel.cpp [hunk 3]
```diff
@@ -166,8 +176,10 @@ winrt::Size ViewPanel::ArrangeOverride(winrt::Size finalSize) {
...
+    float adjustedTop = (float)ViewPanel::GetTop(child) - outerBorderTop;
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/ViewPanel.cpp]:
```suggestion
    float adjustedTop = static_cast<float>(ViewPanel::GetTop(child)) - outerBorderTop;
```

File: vnext/ReactUWP/Views/ViewPanel.cpp [hunk 4]
```diff
@@ -166,8 +176,10 @@ winrt::Size ViewPanel::ArrangeOverride(winrt::Size finalSize) {
...
+
+    child.Arrange(winrt::Rect(adjustedLeft, adjustedTop, (float)childWidth, (float)childHeight));
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/ViewPanel.cpp]:
```suggestion
    child.Arrange(winrt::Rect(adjustedLeft, adjustedTop, static_cast<float>(childWidth), static_cast<float>(childHeight)));
```

**Comment by chrisglein** [comment]
Any thoughts on what we'd need to get some automated testing for this?

**Comment by kmelmon** [comment]
> Any thoughts on what we'd need to get some automated testing for this?

Having tree dumps/master comparisons would be a great way to validate.  It looks like this case also wasn't covered by RNTester so adding something there would be a good first step.

**Comment by ddalp** [comment]
Yeah, I think we should write a simple tree dump native module/viewManager to be able to plug into the e2e test.

---
In reply to: [555262039](https://github.com/microsoft/react-native-windows/pull/3660#issuecomment-555262039) [](ancestors = 555262039)

**Comment by ddalp** [comment]
I will take a look at RNTester too.

---
In reply to: [555262039](https://github.com/microsoft/react-native-windows/pull/3660#issuecomment-555262039) [](ancestors = 555262039)

File: vnext/ReactUWP/Views/ViewPanel.cpp [hunk 5]
```diff
@@ -139,6 +139,16 @@ winrt::Size ViewPanel::MeasureOverride(winrt::Size availableSize) {
 }
 
 winrt::Size ViewPanel::ArrangeOverride(winrt::Size finalSize) {
+  // Sometimes we create outerBorder(i.e. when CornerRadius is true) instead of innerBorder,
+  // Yoga has no notion of outerBorder when calculating the child's position, so we
+  // need to make adujustment in arrange for outerborder's thickness.
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/ViewPanel.cpp]:
```suggestion
  // need to make adjustments in arrange for outerborder's thickness.
```

**Pull request merged**